content/blog/2025-09-15-hash-fields-expiration/index.md

```diff
@@ -0,0 +1,225 @@
+Using a radix tree was possible, but the memory overhead was high. Instead we could use a more lightweight data structure like a skip list, which memory consumption is more bounded
+and is not governed by the data atrophy.
+
+![illustration](hfe-alternative-skiplist.png)
+
+However, This would give `O(log N)` access to the index which did not work well with our target to keep hash operations constant time complexity.
+
+## Coarse Buckets with Adaptive Encoding
+
+Instead of tracking every field’s expiration individually, we designed a coarse bucket system.
+Each field’s timestamp is mapped into a time bucket, represented by a shared “end timestamp.”
+
+![illustration](hfe-coarse-buckets.png)
+
+This solution introduces a semi-sorted data structure which we named 'vset' (stands for "volatile set").
+The vset manages buckets in different time window resolutions and alternating encodings.
+Buckets can split if too many expirations cluster in one interval.
+This adaptability keeps the number of buckets small while ensuring they’re fine-grained enough for efficient cleanup.
+
+When a new field with TTL is added, the vset either places it into an existing bucket or creates a new one.
+If a bucket grows too large, it is narrowed into smaller time windows to keep expiration scans efficient.
+For example: scanning 1,000 items in a single 10-second bucket may result in mostly misses, while spreading them across 10 smaller buckets avoids wasted work.
+Why not always use a very small time window (e.g few milliseconds) for each bucket? The reason is memory.
+Recall that the buckets are managed in a Radix-Tree which introduces a high overhead per each item.
+When many buckets are used, the memory overhead can be high and we might end up with the same memory overhead introduced by using a Radix-Tree index.
+
+The 'vset' also uses different bucket encoding based on the number of items in the bucket.
+A bucket with only a single element would require just a single pointer size of bytes.
+A bucket of small amount of items would be encoded as a vector of item pointers, and when the bucket contains many items we will use the [Valkey hashtable](/blog/new-hash-table) structure to map the relevant items.
+By leveraging data structures we already know how to optimize, we can further improve memory usage and performance with techniques like SIMD acceleration and memory prefetching.
+
+A hash object that contains volatile fields now also carries a secondary vset index.
+At the database level, we maintain a global map of hashes with volatile fields.
+The active expiration cron job scans both regular keys and these hashes, but only iterates over vset buckets whose end time has passed.
+This ensures that CPU time is spent only on fields that are truly ready to expire.
+
+On the storage side, expiration metadata is stored compactly, inlined alongside the hash’s key-value entries.
```

Comment:
I don't know what this sentence means. Is this the RDB format?

Reply:
It just mention how we store the expiration time in the entry